### PR TITLE
Speed up doc2bow

### PR DESCRIPTION
**Pull request by piskvorky**
Continuing from abandoned PR #281 .

### CODE REVIEW

File: gensim/corpora/dictionary.py [hunk 1]
```diff
@@ -145,35 +145,35 @@ def doc2bow(self, document, allow_update=False, return_missing=False):
...
+                    # NOTE this assumes there are no gaps in the id sequence!
+                    token2id[w] = len(token2id)
+
+        result = dict((token2id[w], freq) for w, freq in iteritems(counter)
+                  if w in token2id)
 
         if allow_update:
             self.num_docs += 1
-            self.num_pos += len(document)
+            self.num_pos += sum(itervalues(counter))
```
Comment (piskvorky) [inline gensim/corpora/dictionary.py]: Why this change?

Comment (piskvorky) [inline gensim/corpora/dictionary.py]: Ah never mind, I see now -- Lars' commit introduced a bug where generators no longer worked. Got it.

**Comment by piskvorky** [comment]
I don't understand these pr/push tests... I can't replicate the fails.

Does anyone know what's going on? Nothing changed in the "seed test" AFAIK, not sure why it'd fail now.

CC @cscorley @gojomo

**Comment by cscorley** [comment]
Seems that the test is failing for the "what if this was merged" situation, not the PR itself. Rebasing this PR onto develop's HEAD shows:

```
WARNING:root:LDA failed to converge on attempt 0 (got [[0.33994687274761198, 0.025073027250376752], [0, 0.13983736355699364]], expected [[0.385, 0.022], [0.025, 0.157]])
WARNING:root:LDA failed to converge on attempt 1 (got [[0.33981006363176763, 0.025062020931617739], [0, 0.13988465221995741]], expected [[0.385, 0.022], [0.025, 0.157]])
WARNING:root:LDA failed to converge on attempt 2 (got [[0.3397895136433241, 0.025060365334977817], [0, 0.13989176386377092]], expected [[0.385, 0.022], [0.025, 0.157]])
WARNING:root:LDA failed to converge on attempt 3 (got [[0.33975557414734875, 0.025057635485395614], [0, 0.13990350770285284]], expected [[0.385, 0.022], [0.025, 0.157]])
WARNING:root:LDA failed to converge on attempt 4 (got [[0.339840453569978, 0.025064464773182298], [0, 0.1398741421864324]], expected [[0.385, 0.022], [0.025, 0.157]])
WARNING:root:LDA failed to converge on attempt 5 (got [[0.3397951493428476, 0.025060818994886654], [0, 0.13988981461888891]], expected [[0.385, 0.022], [0.025, 0.157]])
WARNING:root:LDA failed to converge on attempt 6 (got [[0.3397806557230964, 0.025059649107418872], [0, 0.13989483071056127]], expected [[0.385, 0.022], [0.025, 0.157]])
WARNING:root:LDA failed to converge on attempt 7 (got [[0.33978362352577246, 0.025059890173461705], [0, 0.13989380280541175]], expected [[0.385, 0.022], [0.025, 0.157]])
WARNING:root:LDA failed to converge on attempt 8 (got [[0.34003771722484361, 0.025080332380766693], [0, 0.13980600464544757]], expected [[0.385, 0.022], [0.025, 0.157]])
WARNING:root:LDA failed to converge on attempt 9 (got [[0.34030173948506204, 0.025101530899168736], [0, 0.13971506270563669]], expected [[0.385, 0.022], [0.025, 0.157]])
FAIL
```

So, lookup is failing for one of the terms (always=0).

**Comment by cscorley** [comment]
I've changed the PR:

1. Based on head
2. Fixed term lookup with show_topics
3. Adjusted test values

Make sure that item 3 is ok. The values seem to be a lot smoother now, possibly because of changes to the way LDA handles `passes` and `rho_t`

**Comment by piskvorky** [comment]
Thanks Chris. What was the reason for the fails? I still don't get it -- how come the tests worked before?

**Comment by piskvorky** [comment]
Hmm, actually that topic seeding test still fails. I'm not sure what's going on there, will disable it for now.

**Comment by cscorley** [comment]
Strange. Maybe we should seed the RNG for tests that expect exact or close valued tests? The result values aren't off by much from the expected values.

**Comment by piskvorky** [comment]
I don't think so, seeding the tests doesn't make sense. It just covers errors -- better to be explicit. Users won't be picking seeds so that their algorithms work, so neither should tests.
